scheduler/scheduler-server/src/main/java/org/ow2/proactive/scheduler/core/SchedulingMethodImpl.java

```diff
@@ -166,50 +167,102 @@ public int schedule() {
+            LinkedList<EligibleTaskDescriptor> fullListOfTaskRetrievedFromPolicy = currentPolicy.getOrderedTasks(descriptors);
 
-            LinkedList<EligibleTaskDescriptor> taskRetrievedFromPolicy = currentPolicy.getOrderedTasks(descriptors);
+            //if there is no task to scheduled, return without starting any task
+            if (fullListOfTaskRetrievedFromPolicy == null || fullListOfTaskRetrievedFromPolicy.isEmpty()) {
+                return 0;
+            }
+
+            toUnlock = unlockResources(toUnlock);
 
-            //if there is no task to scheduled, return
-            if (taskRetrievedFromPolicy == null || taskRetrievedFromPolicy.isEmpty()) {
-                return numberOfTaskStarted;
+            return getNumberOfTaskStarted(currentPolicy, jobMap, freeResources, fullListOfTaskRetrievedFromPolicy);
+        } finally {
+            if (toUnlock != null) {
+                schedulingService.unlockJobsToSchedule(toUnlock.values());
             }
+        }
+    }
+
+    private int getNumberOfTaskStarted(Policy currentPolicy, Map<JobId, JobDescriptor> jobMap,
+            Set<String> freeResources, LinkedList<EligibleTaskDescriptor> fullListOfTaskRetrievedFromPolicy) {
+        int numberOfTaskStarted = selectAndStartTasks(currentPolicy,
+                                                      jobMap,
+                                                      freeResources,
+                                                      fullListOfTaskRetrievedFromPolicy);
+
+        if (freeResources.isEmpty()) {
+            return numberOfTaskStarted;
+        }
+        if (activeObjectCreationRetryTimeNumber == 0) {
+            return numberOfTaskStarted;
+        }
+
+        return numberOfTaskStarted;
+    }
+
+    private Map<JobId, JobDescriptor> unlockResources(Map<JobId, JobDescriptor> toUnlock) {
+        schedulingService.unlockJobsToSchedule(toUnlock.values());
+        toUnlock = null;
+        return toUnlock;
+    }
+
+    private Set<String> getFreeResources(Policy currentPolicy) {
+        RMState rmState = getRMProxiesManager().getRmProxy().getState();
+        currentPolicy.setRMState(rmState);
+        Set<String> freeResources = rmState.getFreeNodes();
+        if (logger.isDebugEnabled()) {
+            logger.debug("eligible nodes : " + (freeResources.size() < 5 ? freeResources : freeResources.size()));
+        }
+        return freeResources;
+    }
+
+    private void logSelectedJobs(Map<JobId, JobDescriptor> jobMap) {
+        if (logger.isTraceEnabled() && jobMap == null || jobMap.isEmpty()) {
+            logger.trace("No jobs selected to be scheduled");
+        }
+        if (logger.isDebugEnabled() && !jobMap.isEmpty()) {
+            logger.debug("jobs selected to be scheduled : " + (jobMap.size() < 5 ? jobMap : jobMap.size()));
+        }
+    }
+
+    private int selectAndStartTasks(Policy currentPolicy, Map<JobId, JobDescriptor> jobMap, Set<String> freeResources,
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Refactor this method to reduce its Cognitive Complexity from 69 to the 15 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS3776)